libvast/src/table_slice.cpp

```diff
@@ -359,15 +359,22 @@ std::shared_ptr<arrow::RecordBatch> as_record_batch(const table_slice& slice) {
       //                 == table_slice::encoding::arrow) { ... }
       if constexpr (std::decay_t<decltype(*state(encoded, slice.state_))>::encoding
                     == table_slice::encoding::arrow) {
-        return state(encoded, slice.state_)->record_batch();
+        // Get the record batch first, then create a copy that shares the
+        // lifetime with the chunk.
+        auto batch = state(encoded, slice.state_)->record_batch();
+        auto result = std::shared_ptr<arrow::RecordBatch>{
+          batch.get(), [slice, batch](arrow::RecordBatch*) noexcept {
+            VAST_IGNORE_UNUSED(slice);
+            VAST_IGNORE_UNUSED(batch);
+          }};
+        return result;
       } else {
         // Rebuild the slice as an Arrow-encoded table slice.
         auto copy = rebuild(slice, table_slice::encoding::arrow);
         // Bind the lifetime of the copy (and thus the returned Record Batch) to
         // the lifetime of the original slice.
-        slice.chunk_->ref();
-        copy.chunk_->add_deletion_step(
-          [=]() noexcept { slice.chunk_->deref(); });
+        slice.chunk_->add_deletion_step(
+          [copy]() noexcept { VAST_IGNORE_UNUSED(copy); });
```

Comment:
Now that we're actually binding the batch to the copy, we don't need to couple these any more, no?

Reply:
We do, because the record batch lives into the copy.

Calling `as_record_batch` does the following:
1. If called with an an Arrow-encoded table slice, jump to step 5.
2. Convert slice to an Arrow-encoded table slice by making a copy.
3. Bind the lifetime of the copy to the lifetime of the original slice.
4. Call `as_record_batch` with the copy.
5. Get the record batch.
6. Bind the lifetime of the record batch to the slice (and thus to the copy if one exists).
7. Return the record batch.

Reply:
Yes, my hypothesis is that (3) is unnecessary.

Reply:
If you remove (3) then the record batch may be invalid before the table slice is. It's needed so there's a lifetime guarantee both ways. The question is whether we need that.

Reply:
The original slice has no way to access the copy, and neither the other way round. It has nothing to do with the batch.

Reply:
Sounds reasonable. Will remove.